alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs

```diff
@@ -325,3 +332,32 @@ instance
   PrettyA (TxWitness era)
   where
   prettyA = ppTxWitness
+
+instance Era era => Semigroup (Redeemers era) where
```

Comment:
Unfortunately, this Semigroup instance is unlawful and perhaps a bit dangerous. The `Redeemers` type is one of the types that lazily stores its own serialization:
```
newtype Redeemers era = RedeemersConstr (MemoBytes (RedeemersRaw era))
```
So imagine this scenario:
A user out in userland submits a transaction with a redeemer `r`, and they way they choose to cbor encode the redeemer is _not_ what our code would do (but is still legal cbor). Since `r` stores the original bytes of the redeemer, this is what we will use when we need a hash of `r` in the ledger rules. This way the rules and the user can agree to the proper hash of `r`. However, the hash of `r <> mempty` will be different! Why? `r <> mempty` will create a new `Redeemer` using the `Redeemer` pattern, and so the lazy serialization field becomes a thunk. When we later want the hash of `r <> mempty`, we will then serialize the map ourselves, possibly making a different choice for serialization than the user.

Reply:
erm, I guess it's the Monoid instance that has the broken law, but it's `<>` that is causing the grief :)

Reply:
hmm.  The only reason I wrote this instance is to have a "blank" redemer to add witness to.  I was a bit concerned when I wrote the instance.

Reply:
I guess the only valid thing to do is use the pattern constructor at the site I create the redeemer in the elaborator.

Reply:
would a `Default` instance be convenient? We use those in several places, and that won't cause any problems.